### ISSUE TITLE
Gradient checking fails on larger node counts

### ISSUE THREAD

**Issue by ndryden** [LLNL/lbann#588]
Running on Pascal using 16 nodes, gradient checking fails.

```
AL_PROGRESS_RANKS_PER_NUMA_NODE=2 LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/lib64/nvidia_cuda/ OMP_NUM_THREADS=8 srun -N16 --ntasks-per-node=2 --mpibind=off --cpu_bind=mask_cpu:0x000001ff,0x0003fe00 ./build/g
nu.Release.pascal.llnl.gov/install/bin/lbann --model=model_zoo/tests/model_mnist_conv_graph.prototext --reader=model_zoo/data_readers/data_reader_mnist.prototext --optimizer=model_zoo/optimizers/opt_sgd.prototext
...
--------------------------------------------------------------------------------
[3] Epoch : stats formated [tr/v/te] iter/epoch = [1742/194/323]
            global MB = [  31/  31/  31] global last MB = [  29  /  17  /  18  ]
             local MB = [  31/  31/  31]  local last MB = [  29+0/  17+0/  18+0]
--------------------------------------------------------------------------------
Model 0 training epoch 3 objective function : 0.546634
Model 0 training epoch 3 run time : 7.99869s
Model 0 training epoch 3 mini-batch time statistics : 0.00458916s mean, 0.00851253s max, 0.00329945s min, 0.000340482s stdev
Model 0 validation objective function : 11.0021
Model 0 validation run time : 0.285197s
Model 0 validation mini-batch time statistics : 0.00146954s mean, 0.00386422s max, 0.00139175s min, 0.000268499s stdev
--------------------------------------------------------------------------------
Gradient checking...
  Objective function value = 12.9929
  Step size                = 0.00995514
  Expected gradient error  = 0.00157002
Checking branch3_conv_kernel
  GRADIENT ERROR: branch3_conv_kernel, entry (3,0)
    Weight              = 0.059953
    Analytical gradient = -0.00886308
    Numerical gradient  = -0.000894107
    Error               = 0.00796897
    Relative error      = 0.89912
****************************************************************
callback_gradient_check: found large error in gradient
```

I tried going back through our commit history to identify the cause, but reverting to PR merges where @timmoon10 indicates gradient checking was working fine still results in errors for me (e.g. #574). This occurs with and without Aluminum. It works fine on 2, 4, and 8 nodes.

Either we (or a dependency) has a bug that only manifests at larger scales; using floats is too error-prone at larger scale; or our tolerance for gradient checking needs to be node count-dependent.

@timmoon10 thoughts?

**Comment by timmoon10** [comment]
It appears that this caused by local batch norm. At 16 nodes, no process has more than one local mini-batch sample to work with, so I suspect the statistics are getting wonky and putting the learning in a weird place. In any case, gradient checking passes when I use `double` as the datatype. It seems gradients are computed correctly, but the instability is too much for a `float` to handle.
